File: KalkJava/geometrykalk/Cube.java
package geometrykalk;

import java.util.List;


public class Cube extends PlatonicSolid {

	/**
	 * Builds a cube starting from its face
	 */
	protected void createShape(Plane p) throws KalkException {
		//creates eight points representing the vertices of the cube (in the first quadrant)
		if(p instanceof Square) {
			Square s = (Square) p;
	        double h = s.side();
	        List<Point2> vP = s.getPoints();
	        for(Point2 point : vP) {
	            setPoint(point.getX(), point.getY(), 0);
	            setPoint(point.getX(), point.getY(), h);
	        }
	    } else throw new KalkException("It is not possible to build the object from the current plane shape.");
	}

	
	/**
	 * Constructor with one double type parameter
	 * @param s represents the length of the edge
	 * @throws KalkException if r is not a positive value
	 */
	public Cube(double e) throws KalkException {
		super("Cube");
		Square s = new Square(e);
		createShape(s);
	}
	
	
	/**
	 * Default constructor
	 */
	public Cube() throws KalkException {
		this(1);
	}
	
	
	/**
	 * Constructor with one parameter, which represents its face.
	 * Used for 2D->3D conversion
	 */
	public Cube(Square s) throws KalkException {
		super("Cube");
		createShape(s);
	}

	
	/**
	 * @return the length of the space diagonal, which is the segment connecting two vertices that are not on the same face
	 */
	public double spaceDiagonal() {
	    return edge() * Math.sqrt(3);
	}
	
	
	/**
	 * @return the surface of the solid, excluding the area of its base and top faces
	 */
	public double lateralSurface() {
	    return 4 * Math.pow(edge(), 2);
	}
	
	
	public int schlafliSymbol_q() {
		return 3;
	}
	
	
	public double inRadius() {
	    return edge() / 2;
	}
	
	
	public double circumRadius() {
	    return edge() * Math.sqrt(3) / 2;
	}
	
	
	public double surface() {
	    return 6 * Math.pow(edge(), 2);

	}
	
	
	public double volume() {
	    return Math.pow(edge(), 3);

	}
	

	public Square face() throws KalkException {
	    return new Square(edge());
	}

	
	public Point3 centreOfSymmetry() throws KalkException {
		Point2 p = face().centreOfSymmetry();
	    return new Point3(p.getX(), p.getY(), edge() / 2);
	}

	
	public void expand(double d) throws KalkException {
		if(d < 0)
			throw new KalkException("Negative values not allowed.");
		Cube c = new Cube(edge() + edge() * d / 100);
	    this.setEqual(c);
	}
	
	
	public void reduce(double d) throws KalkException {
		if(d < 0 || d >= 100)
			throw new KalkException("Value out of range. It must be between 0 and 100.");
		Cube c = new Cube(edge() + edge() * d / 100);
	    this.equals(c);
	}
	

	public Cube sum(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Cube c = (Cube) s;
	    double newEdge = edge() + c.edge();
	    return new Cube(newEdge);
	}
	
	
	public Cube difference(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Cube c = (Cube) s;
	    Square faceDifference = face().difference(c.face()); //it may throw exceptions
	    return new Cube(faceDifference);
	}
	
	
	public Cube multiplication(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Cube c = (Cube) s;
	    double newEdge = edge() * c.edge();
	    return new Cube(newEdge);
	}
	
	
	public Cube division(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Cube c = (Cube) s;

	    //both operands are greater than 0
	    double newEdge = edge() / c.edge();
	    return new Cube(newEdge);
	}
}

File: Figure/Model/cube.cpp
#include "cube.h"
#include <math.h>
#include <typeinfo>


Cube::Cube(double e) : PlatonicSolid("Cube") {
    Square* pS = new Square(e); //it may throw an exception
    createShape(pS);
    delete pS;
}


Cube::Cube(Square* pS) : PlatonicSolid("Cube") {
    createShape(pS);
}


void Cube::createShape(const Plane* p) {
    //creates eight points representing the vertices of the cube (in the first quadrant)
    const Square* pS = dynamic_cast<const Square*>(p);
    if(pS) {
        double h = pS->side();
        const std::vector<Point2> vP = pS->getPoints();
        for(std::vector<Point2>::const_iterator it = vP.begin(); it != vP.end(); ++it) {
            setPoint(it->getX(), it->getY(), 0);
            setPoint(it->getX(), it->getY(), h);
        }
    } else throw WrongShape();
}


double Cube::spaceDiagonal() const{
    return edge() * sqrt(3);
}


double Cube::lateralSurface() const {
    return 4 * pow(edge(), 2);
}


int Cube::schlafliSymbol_q() const { return 3; }


double Cube::inRadius() const {
    return edge() / 2;
}


double Cube::circumRadius() const {
    return edge() * sqrt(3) / 2;
}


double Cube::surface() const {
    return 6 * pow(edge(), 2);
}


double Cube::volume() const {
    return pow(edge(), 3);
}


Square* Cube::face() const {
    return new Square(edge());
}


Point3* Cube::centreOfSymmetry() const {
    Point2* pP = face()->centreOfSymmetry();
    return new Point3(pP->getX(), pP->getY(), edge() / 2);
}


void Cube::expand(double n) {
    if(n < 0) throw NegativeValue();
    Cube* pC = new Cube(edge() + edge() * n / 100);
    *this = *pC;
    delete pC;
}


void Cube::reduce(double n) {
    if(n < 0 || n >= 100) throw ValueOutOfRange();
    Cube* pC = new Cube(edge() - edge() * n / 100);
    *this = *pC;
    delete pC;
}


Cube* Cube::operator+(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Cube& rC = static_cast<const Cube&>(s);
    double newEdge = edge() + rC.edge();
    return new Cube(newEdge);
}


Cube* Cube::operator-(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Cube& rC = static_cast<const Cube&>(s);
    Square* pS = *face() - *(rC.face()); //it may throw exceptions
    return new Cube(pS);
}


Cube* Cube::operator*(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Cube& rC = static_cast<const Cube&>(s);
    double newEdge = edge() * rC.edge();
    return new Cube(newEdge);
}


Cube* Cube::operator/(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Cube& rC = static_cast<const Cube&>(s);

    //both operands are greater than 0
    double newEdge = edge() / rC.edge();
    return new Cube(newEdge);
}
